lib/engine/game/g_18_fl/step/token.rb

```diff
@@ -42,16 +48,36 @@ def process_hex_token(action)
             token = action.token
 
             raise GameError, "#{hex.id} is not a town" if hex.tile.towns.empty?
-            raise GameError, "#{entity.name} already has a hotel in #{hex.tile.location_name}" if tokened(hex,
-                                                                                                          entity)
+            raise GameError, "#{entity.name} already has a hotel in #{hex.tile.location_name}" if tokened(hex, entity)
+
+            cost = action.cost # We are using token_cost_override
+            raise GameError, "#{entity.name} cannot afford "\
+                  "#{@game.format_currency(cost)} cost to lay hotel" if cost > entity.cash
 
-            @game.log << "#{entity.name} places a hotel on #{hex.name}"
+            @game.log << "#{entity.name} places a hotel on #{hex.name} for #{@game.format_currency(cost)}"
+            entity.spend(cost, @game.bank)
 
             entity.tokens.delete(token)
             hex.tile.icons << Part::Icon.new("../logos/18_fl/#{entity.id}")
             pass!
           end
 
+          def token_cost_override(entity, city_hex, _slot, token, _tokener)
```

Comment:
is this only run once ever? does it use auto actions  or is it not calculated when cost is set?

Reply:
It is run once when creating the action in the view

Reply:
I made it so the step verifies the cost (only when not @loading) so the server can verify the cost
This increases it to running twice total, but it's only when clicking to place a token